src/Microsoft.PowerShell.Commands.Utility/commands/utility/NewTemporaryFileCommand.cs

```diff
@@ -12,18 +12,57 @@ namespace Microsoft.PowerShell.Commands
     [OutputType(typeof(System.IO.FileInfo))]
     public class NewTemporaryFileCommand : Cmdlet
     {
+        private string _extension = ".tmp";
+
+        /// <summary>
+        /// Specify a different file extension other than the default one, which is '.tmp'. The period in this parameter is optional
+        /// </summary>
+        [Parameter(Position = 0)]
+        [ValidateNotNullOrEmpty]
+        public string Extension
+        {
+        get { return _extension; }
+        set { _extension = value; }
+        }
+
         /// <summary>
         /// returns a TemporaryFile
         /// </summary>
         protected override void EndProcessing()
         {
             string filePath = null;
-            string tempPath = System.Environment.GetEnvironmentVariable("TEMP");
-            if (ShouldProcess(tempPath))
+            string tempPath = Path.GetTempPath();
+            int attempts = 0;
+            bool temporaryFilePathFound = false;
+            while(attempts++ < 10 && !temporaryFilePathFound)
+            {
+                string fileName = Path.GetRandomFileName();
+                fileName = Path.ChangeExtension(fileName, _extension);
+                filePath = Path.Combine(tempPath, fileName);
+                if(!File.Exists(filePath))
+                {
+                    temporaryFilePathFound = true;
+                }
+            }
+            if(!temporaryFilePathFound)
+            {
+                WriteError(
+                    new ErrorRecord(
+                            new IOException("Could not find an available temporary file name in {tempPath}."),
+                            "NewTemporaryFileResourceUnavailable",
+                            ErrorCategory.ResourceUnavailable,
+                            tempPath));
+                    return;
+            }
+
+            if (ShouldProcess(filePath))
```

Comment:
While this implicitly fixes the bug in the original code that referred to environment variable `TEMP`, which can only be assumed to exist in Windows, note that you've changed the logic from reporting the _directory_ path to the _file_ path.

In other words: to remain consistent with how thing worked before, it should remain `if (ShouldProcess(tempPath))`

I guess the logic is: `New-TemporaryFile` operates on the temp. _directory_ as its target, and creates a new file _inside it_.

```
What if: Performing the operation "New-TemporaryFile" on target "C:\Users\jdoe\AppData\Local\Temp".
```